jsonutil/puritan/puritan.go

```diff
@@ -0,0 +1,340 @@
+	}
+	return rune(u)
+}
+
+type uOpt struct {
+	nlok  bool
+	basic bool
+}
+
+func unmarshal(in []byte, o uOpt) (string, error) {
+	// heavily based on (inspired by?) unquoteBytes from encoding/json
+
+	if len(in) < 2 || in[0] != '"' || in[len(in)-1] != '"' {
+		// maybe it's a null and that's alright
+		if len(in) == 4 && in[0] == 'n' && in[1] == 'u' && in[2] == 'l' && in[3] == 'l' {
+			return "", nil
+		}
+		return "", fmt.Errorf("missing string delimiters: %q", in)
+	}
+
+	// prune the quotes
+	in = in[1 : len(in)-1]
+	i := 0
+	// try the fast track
+	for i < len(in) {
+		// 0x00..0x19 is the first of Cc
+		// 0x20..0x7e is all of printable ASCII (minus control chars)
+		if in[i] < 0x20 || in[i] > 0x7e || in[i] == '\\' || in[i] == '"' {
+			break
+		}
+		i++
+	}
+	if i == len(in) {
+		// wee
+		return string(in), nil
+	}
+	if o.basic {
+		return "", fmt.Errorf("invalid simple JSON string %q: not simple at %d", in, i)
+	}
+	// in[i] is the first problematic one
+	out := make([]byte, i, len(in)+2*utf8.UTFMax)
+	copy(out, in)
+	var r, r2 rune
+	var sz int
+	var c byte
+	var ubuf [utf8.UTFMax]byte
+	for i < len(in) {
+		switch c = in[i]; {
+		case c == '"':
+			return "", fmt.Errorf("unexpected unescaped quote at %d in \"%s\"", i, in)
+		case c < 0x20:
+			return "", fmt.Errorf("unexpected control character at %d in %q", i, in)
+		case c == '\\':
+			// handle escapes
+			i++
+			if i == len(in) {
+				return "", fmt.Errorf("unexpected end of string (trailing backslash) in \"%s\"", in)
+			}
+			switch in[i] {
+			case 'u':
+				// oh dear, a unicode wotsit
+				r = u4(in[i-1:])
+				if r < 0 {
+					x := in[i-1:]
+					if len(x) > 6 {
+						x = x[:6]
+					}
+					return "", fmt.Errorf(`badly formed \u escape %q at %d of "%s"`, x, i, in)
+				}
+				i += 5
+				if utf16.IsSurrogate(r) {
+					// sigh
+					r2 = u4(in[i:])
+					if r2 < 0 {
+						x := in[i:]
+						if len(x) > 6 {
+							x = x[:6]
+						}
+						return "", fmt.Errorf(`badly formed \u escape %q at %d of "%s"`, x, i, in)
+					}
+					i += 6
+					r = utf16.DecodeRune(r, r2)
+				}
+				if r <= 0x9f {
+					// otherwise, it's Cc (both halves, as we're looking at runes)
+					if (o.nlok && r == '\n') || (r >= 0x20 && r <= 0x7e) {
+						out = append(out, byte(r))
+					}
+				} else if r != unicode.ReplacementChar && !unicode.Is(strutil.Ctrl, r) {
+					sz = utf8.EncodeRune(ubuf[:], r)
+					out = append(out, ubuf[:sz]...)
+				}
+			case 'b', 'f', 'r', 't':
+				// do nothing
+				i++
+			case 'n':
+				if o.nlok {
+					out = append(out, '\n')
+				}
+				i++
+			case '"', '/', '\\':
+				// the spec says just ", / and \ can be backslash-escaped
+				// but go adds ' to the list (why?!)
+				out = append(out, in[i])
+				i++
+			default:
+				return "", fmt.Errorf(`unknown escape '%c' at %d of "%s"`, in[i], i, in)
+			}
+		case c <= 0x7e:
+			// printable ASCII, except " or \
+			out = append(out, c)
+			i++
+		default:
+			r, sz = utf8.DecodeRune(in[i:])
+			j := i + sz
```

Comment:
That might be nicer as "size".. or just "n".. sz tastes like something else, even more in the context of unicode which has pairs of chars for many unclear purposes. There's at least one more entry above as well.